src/transform/src/analysis/equivalences.rs

```diff
@@ -0,0 +1,582 @@
+                // TODO: COUNT ensures a non-null value.
+                equivalences.project(input_arity..(input_arity + group_key.len()));
+                equivalences
+            }
+            MirRelationExpr::TopK { .. } => results.last().unwrap().clone(),
+            MirRelationExpr::Negate { .. } => results.last().unwrap().clone(),
+            MirRelationExpr::Threshold { .. } => results.last().unwrap().clone(),
+            MirRelationExpr::Union { .. } => {
+                // TODO: `EquivalenceClasses::union` takes references, and we could probably skip much of this cloning.
+                let expr_index = results.len();
+                depends
+                    .children_of_rev(expr_index, expr.children().count())
+                    .map(|c| results[c].clone())
+                    .reduce(|e1, e2| e1.union(&e2))
+                    .unwrap()
+            }
+            MirRelationExpr::ArrangeBy { .. } => results.last().unwrap().clone(),
+        };
+
+        let expr_type = &depends.results::<RelationType>().unwrap()[results.len()];
+        equivalences.minimize(expr_type);
+        equivalences
+    }
+}
+
+/// A compact representation of classes of expressions that must be equivalent.
+///
+/// Each "class" contains a list of expressions, each of which must be `Eq::eq` equal.
+/// Ideally, the first element is the "simplest", e.g. a literal or column reference,
+/// and any other element of that list can be replaced by it.
+///
+/// The classes are meant to be minimized, with each expression as reduced as it can be,
+/// and all classes sharing an element merged.
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Default, Debug)]
+pub struct EquivalenceClasses {
+    /// Multiple lists of equivalent expressions, each representing an equivalence class.
+    ///
+    /// The first element should be the "canonical" simplest element, that any other element
+    /// can be replaced by.
+    /// These classes are unified whenever possible, to minimize the number of classes.
+    pub classes: Vec<Vec<MirScalarExpr>>,
+}
+
+impl EquivalenceClasses {
+    /// Sorts and deduplicates each class, and the classes themselves.
+    fn tidy(&mut self) {
+        for class in self.classes.iter_mut() {
+            // Remove all literal errors, as they cannot be equated to other things.
+            class.retain(|e| !e.is_literal_err());
+            class.sort_by(|e1, e2| match (e1, e2) {
+                (MirScalarExpr::Literal(_, _), MirScalarExpr::Literal(_, _)) => e1.cmp(e2),
+                (MirScalarExpr::Literal(_, _), _) => std::cmp::Ordering::Less,
+                (_, MirScalarExpr::Literal(_, _)) => std::cmp::Ordering::Greater,
+                (MirScalarExpr::Column(_), MirScalarExpr::Column(_)) => e1.cmp(e2),
+                (MirScalarExpr::Column(_), _) => std::cmp::Ordering::Less,
+                (_, MirScalarExpr::Column(_)) => std::cmp::Ordering::Greater,
```

Comment:
Could you add a comment here mentioning that the first elements of the classes are the representatives, and these ordering considerations are important because we'll be simplifying expressions to these representatives?